### PR TITLE
ART-3003 Promote and Multi promote enhancements

### PR DESCRIPTION
**Pull request by thegreyd**
- Have a "in_flight_prev" param on promote job which if specified correctly does not prompt artist for input.
- Have multi_promote job include params
    - "resume_from" 
    - "in_flight_prev" 

Test job runs
- https://saml.buildvm.openshift.eng.bos.redhat.com:8888/job/hack/job/sidsharm-aos-cd-jobs/job/build%252Fpromote/34/parameters/ (with in_flight_prev)
- https://saml.buildvm.openshift.eng.bos.redhat.com:8888/job/hack/job/sidsharm-aos-cd-jobs/job/build%252Fmulti_promote/ (with in_flight_prev and selective nightlies with confirmation prompt)

### CODE REVIEW

File: jobs/build/promote/Jenkinsfile [hunk 1]
```diff
@@ -327,25 +333,42 @@ node {
...
-                                    defaultValue: "${suggest_previous}",
-                                    description: (acquire_failure?acquire_failure:"Doozer thinks these are the other releases to include.") + " Edit as necessary (comma delimited).",
-                                    name: 'SUGGESTED',
-                                ),
-                            ]
-                        )
+                    in_flight_prev_required = true
+                    if (params.IN_FLIGHT_PREV) {
+                        in_flight_prev = params.IN_FLIGHT_PREV.trim()
+                        pattern = /4\.$prevMinor\.(\d+)/
```
Comment (vfreex) [inline jobs/build/promote/Jenkinsfile:339]: We can eliminate the hardcoded `4` major version by using the `major` variable defined at line 210.

File: jobs/build/promote/Jenkinsfile [hunk 2]
```diff
@@ -327,25 +333,42 @@ node {
...
+                                        description: (acquire_failure?acquire_failure:"Doozer thinks these are the other releases to include.") + " Edit as necessary (comma delimited).",
+                                        name: 'SUGGESTED',
+                                    ),
+                                ]
+                            )    
+                        }
+                        in_flight_prev = resp.IN_FLIGHT_PREV
                     }
+                    
+                    previousList = commonlib.parseList(resp.SUGGESTED) + commonlib.parseList(in_flight_prev)
```
Comment (vfreex) [inline jobs/build/promote/Jenkinsfile:371]: If `in_flight_prev_required` is false, `resp.SUGGESTED` here will became undefined. I think we still need to promote the artist for the suggested previous unless we have a good way to get rid of that.

File: jobs/build/promote/Jenkinsfile [hunk 3]
```diff
@@ -327,25 +333,42 @@ node {
...
-                                ),
-                            ]
-                        )
+                    in_flight_prev_required = true
+                    if (params.IN_FLIGHT_PREV) {
+                        in_flight_prev = params.IN_FLIGHT_PREV.trim()
+                        pattern = /4\.$prevMinor\.(\d+)/
+                        match = in_flight_prev =~ pattern
+                        match.find()
+                        if (match.size() == 1) {
```
Comment (vfreex) [inline jobs/build/promote/Jenkinsfile:342]: Would also like a way to specify an empty `in_flight_prev`. e.g. currently we always leave that input box empty if no 4.5.z release in flight.

**Review by vfreex — CHANGES_REQUESTED** [review]
I like the idea to have less prompts, but seems there are some issues. See my comments inline.

File: jobs/build/promote/Jenkinsfile (hunk 1 above)
Comment (thegreyd) [reply jobs/build/promote/Jenkinsfile]: Done

File: jobs/build/promote/Jenkinsfile (hunk 2 above)
Comment (thegreyd) [reply jobs/build/promote/Jenkinsfile]: Thanks for catching that, fixed it

File: jobs/build/promote/Jenkinsfile (hunk 3 above)
Comment (thegreyd) [reply jobs/build/promote/Jenkinsfile]: added an option of "skip"

**Review by vfreex — APPROVED** [review]
Thanks.
Basically LGTM. Would like to try this in this week's release preparation.

/lgtm

**Pull request merged**